File: trunk/math/matrix.cpp
template <typename T>
class Matrix {
 public:
  Matrix(int row, int col, const T* arr = NULL, const T& mod = T(0))
      : _row(row), _col(col), _mat(new T[row * col]), _mod(mod) {
    for (int i = 0; i < _row; i++) {
      for (int j = 0; j < _col; j++) {
        if (arr == NULL) {
          (*this)[i][j] = 0;
        } else {
          (*this)[i][j] = arr[i * col + j];
        }
      }
    }
  }

  Matrix(const Matrix<T>& other)
      : _row(0), _col(0), _mat(NULL) { *this = other; }

  void operator =(const Matrix<T>& other) {
    if (_mat == NULL || _row * _col != other.row() * other.col()) {
      if (_mat != NULL) delete [] _mat;
      _mat = new T[other.row() * other.col()];
    }

    _row = other.row();
    _col = other.col();
    _mod = other.mod();

    for (int i = 0; i < _row; i++) {
      for (int j = 0; j < _col; j++) {
        (*this)[i][j] = other[i][j];
      }
    }
  }

  ~Matrix() { delete [] _mat; }

  int row() const { return _row; }

  int col() const { return _col; }

  T mod() const { return _mod; }

  T* operator [](int row) const {
    return _mat + row * _col;
  }

  Matrix<T> operator *(const Matrix& other) const {
    assert(_col == other.row());

    Matrix<T> prod(_row, other.col(), NULL, _mod);

    for (int i = 0; i < _row; i++) {
      for (int j = 0; j < other.col(); j++) {
        for (int k = 0; k < _col; k++) {
          prod[i][j] += (*this)[i][k] * other[k][j];
          if (_mod != 0) { prod[i][j] %= _mod; }
        }
      }
    }
    return prod;
  }

  Matrix<T> pow(long long k) const {
    k--;
    Matrix<T> pd(*this);
    Matrix<T> pw(*this);
    while (k > 0) {
      if (k & 1) pd = pd * pw;
      if (k > 1) pw = pw * pw;
      k /= 2;
    }
    return pd;
  }

 private:
  int _row;
  int _col;
  T* _mat;
  T _mod;
};

File: math/matrix.py
class Matrix(object):

    def __init__(self, row_size, col_size, data):
        self.row_size = row_size
        self.col_size = col_size
        self.data = data[:]

    def __mul__(self, other):
        row_size, col_size = self.row_size, other.col_size
        product = Matrix(row_size, col_size, [0]*row_size*col_size)
        for i in xrange(row_size):
            for j in xrange(col_size):
                for k in xrange(self.col_size):
                    product.data[col_size*i + j] += self.get(i, k) *  other.get(k, j)
        return product

    def __pow__(self, k, mod):
        from copy import deepcopy
        pd, pw = deepcopy(self), self
        k -= 1
        while k > 0:
            if k % 2 == 1:
                pd = pd.mulmod(pw, mod)
            if k > 1:
                pw = pw.mulmod(pw, mod)
            k /= 2
        return pd

    def __repr__(self):
        lines = []
        for i in xrange(self.row_size):
            lines.append(' '.join(map(str, self.data[self.col_size*i:self.col_size*(i+1)])))
        return '\n'.join(lines)

    def mulmod(self, other, mod):
        pd = self * other
        for i in xrange(len(pd.data)):
            pd.data[i] %= mod
        return pd

    def get(self, row, col):
        return self.data[self.col_size*row + col]

